articles/quickstart/native/ionic4/01-login.md

```diff
@@ -0,0 +1,443 @@
+    public zone: NgZone,
+    private storage: Storage,
+    private safariViewController: SafariViewController
+  ) {
+    this.storage.get('profile').then(user => this.user = user);
+    this.storage.get('access_token').then(token => this.accessToken = token);
+    this.storage.get('expires_at').then(exp => {
+      this.loggedIn = Date.now() < JSON.parse(exp);
+      this.loading = false;
+    });
+  }
+
+  login() {
+    this.loading = true;
+    const options = {
+      scope: 'openid profile offline_access'
+    };
+    // Authorize login request with Auth0: open login page and get auth results
+    this.Client.authorize(options, (err, authResult) => {
+      if (err) {
+        throw err;
+      }
+      // Set access token
+      this.storage.set('access_token', authResult.accessToken);
+      this.accessToken = authResult.accessToken;
+      // Set access token expiration
+      const expiresAt = JSON.stringify((authResult.expiresIn * 1000) + new Date().getTime());
+      this.storage.set('expires_at', expiresAt);
+      // Set logged in
+      this.loading = false;
+      this.loggedIn = true;
+      // Fetch user's profile info
+      this.Auth0.client.userInfo(this.accessToken, (err, profile) => {
+        if (err) {
+          throw err;
+        }
+        this.storage.set('profile', profile).then(val =>
+          this.zone.run(() => this.user = profile)
+        );
+      });
+    });
+  }
+
+  logout() {
+    this.storage.remove('profile');
+    this.storage.remove('access_token');
+    this.storage.remove('expires_at');
+    this.accessToken = null;
+    this.user = null;
+    this.loggedIn = false;
+
+    this.safariViewController.isAvailable()
+      .then((available: boolean) => {
+        const domain = AUTH_CONFIG.domain;
+        const clientId = AUTH_CONFIG.clientId;
+        const pkgId = AUTH_CONFIG.packageIdentifier;
+        let url = `https://<%= "${domain}" %>/v2/logout?client_id=<%= "${clientId}" %>&returnTo=<%= "${pkgId}" %>://<%= "${domain}" %>/cordova/<%= "${pkgId}" %>/callback`;
+        if (available) {
+          this.safariViewController.show({
+            url: url
+          })
+          .subscribe((result: any) => {
+              if(result.event === 'opened') console.log('Opened');
+              else if(result.event === 'loaded') console.log('Loaded');
+              else if(result.event === 'closed') console.log('Closed');
+            },
+            (error: any) => console.error(error)
+          );
+        } else {
+          // use fallback browser
+          cordova.InAppBrowser.open(url, '_system');
+        }
+      }
+    );
+  }
+}
+```
+
+## Add Authentication with Auth0
+
+Add a control to your app to allow users to log in. The control should call the `login` method from the `AuthService`. Start by injecting the `AuthService` in a component.
+
+```js
+// src/app/home/home.page.ts
+import { Component } from '@angular/core';
+import { AuthService } from '../../services/auth.service';
```

Comment:
I think this is going back too far, I think it needs to be `../services/auth.service`

Reply:
I think previously you had it correct with `../services/auth.service` and I had asked you to change it, not sure why...